### PR TITLE
Add opt-in snapshot_steps to bound memory use of the time-history output

### PR DESCRIPTION
**Pull request by zhiping0913**
## Why

`_simulation`'s `lax.scan` stacks `positions`, `velocities`, `electric_field`,
`magnetic_field`, `current_density`, and `charge_density` for **every one**
of `total_steps` iterations (that's what `lax.scan`'s `ys` output does by
default). For large grids, large particle counts, or long runs, this
per-step history is what actually exhausts GPU memory — even when the
caller only cares about the final state, or a handful of snapshots along
the way.

## How to use it

A new, opt-in `solver_parameters["snapshot_steps"]` lets you pass an
explicit list of step indices to keep. Everything else about `Simulation`
is unchanged.

```python
# Only the final state:
"solver_parameters": {
    "snapshot_steps": [total_steps - 1],
},

# Evenly spaced snapshots:
"solver_parameters": {
    "snapshot_steps": np.arange(total_steps)[::save_every],
},

# A few arbitrary specific times:
"solver_parameters": {
    "snapshot_steps": [0, 100, 200, total_steps - 1],
},
```

Left unset (the default), behavior is **exactly** what it was before this
PR: full per-step history via `lax.scan`'s normal stacking. Nothing changes
for existing code/configs that don't set this key.

`time_array` is realigned to whichever steps you asked for, so
`output["time_array"][i]` always matches `output["electric_field"][i]`
(etc.) regardless of mode.

### Implementation

When `snapshot_steps` is set, instead of letting `lax.scan` stack a
per-step output, a small fixed-size buffer (`jax.eval_shape`-sized from one
sample step) is carried through the scan and written into via `lax.cond` +
`.at[idx].set(...)` only on the requested steps; the scan's `ys` output is
`None`, so nothing gets stacked. Memory then scales with
`len(snapshot_steps)` instead of `total_steps`.

## Effect (measured)

Small case (`N=2000` grid points, `total_steps=555`, laser-pulse setup from
`test_snapshot_steps.py` with `CELLS_PER_LAMBDA=100`):

| mode | `electric_field.shape` | peak GPU memory |
|---|---|---|
| unset (full history) | `(555, 2000, 3)` | 122.3 MB |
| `snapshot_steps=[554]` | `(1, 2000, 3)` | 96.2 MB |
| | | **-21.4%** |

Same setup at `CELLS_PER_LAMBDA=1000` (`N=20000`, `total_steps=5555`) —
representative of the scale that actually OOMs:

| mode | `electric_field.shape` | peak GPU memory |
|---|---|---|
| unset (full history) | `(5555, 20000, 3)` | 8540.3 MB |
| `snapshot_steps=[5554]` | `(1, 20000, 3)` | 6106.6 MB |
| | | **-28.5% (-2.4 GB)** |

Both measured with `jax.devices()[0].memory_stats()["peak_bytes_in_use"]`
(`XLA_PYTHON_CLIENT_PREALLOCATE=false`, each mode in its own subprocess so
the peak — a monotonically non-decreasing high-water mark — isn't
contaminated by the other run). The savings scale with
`(total_steps - len(snapshot_steps)) × per-step array size`, so they grow
essentially linearly with grid size and run length; for a large production
run (bigger grid, more steps) the absolute reduction would be proportionally
larger than either of these two examples.

## Testing

Added `tests/test_snapshot_steps.py`:

- `test_snapshot_steps_matches_full_history_final_state` — always runs
  (CPU-safe). Confirms output shapes shrink from `(total_steps, ...)` to
  `(len(snapshot_steps), ...)`, `time_array` lines up with the requested
  steps, and every recorded snapshot exactly matches the corresponding step
  of an equivalent full-history run with the same seed.
- `test_snapshot_steps_reduces_peak_gpu_memory` — runs a larger problem
  twice in isolated subprocesses (full history vs. final-step-only) and
  asserts snapshot-mode peak GPU memory is strictly lower. Skipped
  automatically when no GPU backend is available (e.g. on this repo's
  CPU-only CI runners), since `peak_bytes_in_use` isn't a meaningful
  comparison on CPU.

Full `tests/test_simulation.py` + `tests/test_snapshot_steps.py` suite
(18 tests) passes locally with no regressions.

## Environment used for testing

- `jaxincell` `0.2.dev17+g5268a1556.d20260716` (built from this branch, on top of `5268a15`)
- `jax` `0.10.0`
- Python 3.12.13
- GPU: 2× NVIDIA A100

### CODE REVIEW

**Comment by dseidita1919** [comment]
This is a very valuable addition to the code. There are a few changes I would like to see before we merge it, though.

First, this new parameter needs support in the _parameters portion of the code. It should be added to DEFAULT_SOLVER_PARAMETERS along with its default value of None and support should be added in the clean_and_initialize_solver_parameters function to ensure functionality is properly retained for any usage of the code including importing via toml.

It also feels prudent to add support for integer values of snapshot_steps. The clean_and_initialize_solver_parameters function can check if it is an integer and if it is, set it to an array with a number of equally spaced snapshots equal to the integer. Essentially,
`snapshot_steps = 5`
becomes
`snapshot_steps = [ 0, 5, 10, 15, 20 ]`
for `total_steps = 20`
Doing this allows users to still have the ability to set equally spaced arrays without requiring them to write code beyond the input file.

Additionally, there is a lot of split logic created by this PR. Specifically, this involves lines 257 (set snapshot_step_list), 266 (change logic between not using snapshot_steps and using snapshot_steps), and line 314 (correctly determine the time_array). These three splits should be combined into only one logical split. This will make the code more readable and keep the throughline of the code more consistent meaning there are less opportunities for logical splits to lead to bugs or differences in activity depending on specific usage. The main loop logic can be fixed to use the buffer system introduced in this PR always instead of just when using snapshots which will remove the middle logical split. Then, the snapshot_step_list can be built regardless of running with snapshot_steps or not, i.e. running without snapshot_steps sets snapshot_steps to be every step. These changes will make the throughline of the code as consistent as possible and remove the most number of conditionals from the jit'ed portion of the code.

If you would like, @zhiping0913 , I can work on some of these changes myself. I'd like to hear from you before making any changes involving your branch, however.

**Comment by zhiping0913** [comment]
Hi,
Thank you for your review and reply!
The purpose of my proposed change is to enable the program to retain only a portion of the snapshots instead of all of them, thereby reducing the pressure on memory, especially in scenarios where there are a considerable number of simulation steps.
It is a good idea to combine them into one logic and force the program to always use the buffer system!
The only thing is that I prefer using snapshot_steps_every=5 instead of snapshot_steps = 5 to indicate the equally spaced snapshots to avoid ambulation.
I would be very grateful if you could work on some of these changes yourself and make the code consistent. I could think about more examples to test whether it works and whether these changes can make the end-to-end automatic differentiation more efficient.
Best

**Comment by dseidita1919** [comment]
Made some changes to tighten up the logic and add support and cleaning in _solver_parameters. I did not look at integer value inputs or adding snapshot_steps_every at this point.

Another thing worth thinking about is dealing with a situation where the last step of the simulation is not included in snapshot_steps. Should the simulation cut off short? Should the last step of the simulation automatically be added to snapshot_steps? Something to think about.

Tests added should see double checking - made with Codex and I did not validate them especially thoroughly.

**Comment by zhiping0913** [comment]
I think we can include 0 and N_step in snapshot_steps by default
